### PR TITLE
AS COMING

### PR DESCRIPTION
**Pull request by Graph-Donte-Crypto**

### CODE REVIEW

**Comment by Graph-Donte-Crypto** [comment]
NONOGRAM-GAMER-RS 2024-02-11 10-56-42

File: src/ai_player.rs [hunk 1]
```diff
@@ -0,0 +1,56 @@
+use crate::description::LevelDescription;
+use crate::game_state::GameGridState;
+use crate::grid::Grid;
+use crate::play_engine::PlayEngine;
+use std::time::{Duration, Instant};
+
+pub struct AiPlayer {
+    pub engines: Vec<Box<dyn PlayEngine>>,
```
Comment (alexesmet) [inline src/ai_player.rs]: PlayEngine -> SolvingStrategy ?

File: src/transaction.rs [hunk 2]
```diff
@@ -21,6 +21,22 @@ impl Transaction {
         self.changes.push(TransactionDetails {col, row, old_state, new_state})
     }
 }
+impl Clone for Transaction {
```
Comment (alexesmet) [inline src/transaction.rs]: Can be removed by putting `#[derive(Clone)]` on the `Transaction` and `TransactionDetails`

File: src/main.rs [hunk 3]
```diff
@@ -316,7 +436,52 @@ impl EventHandler for MyGame {
             }
         }
 
+        for row_description in self.game_state.lvl_desc().rows.iter().enumerate() {
+            for row_description_part in row_description.1.iter().enumerate() {
```
Comment (alexesmet) [inline src/main.rs:443]: :exclamation:  Use tuple destructuring to define variable names
```rust
            for (row_id, row_item) in row_description.1.iter().enumerate() {
```

Comment (alexesmet) [reply src/main.rs:443]: :nerd_face:

File: src/simple_engine.rs [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+use crate::game_state::CellState;
+use crate::grid::Grid;
+use crate::play_engine::PlayEngine;
+
+pub struct SimpleEngine {
+
+}
+impl PlayEngine for SimpleEngine {
+    fn play_one_turn(&self, level_description: &LevelDescription, grid: &mut dyn Grid) -> bool {
+        for row_description in level_description.rows.iter().enumerate() {
```
Comment (alexesmet) [inline src/simple_engine.rs]: :exclamation: Use tuple destruct.

File: src/simple_engine.rs [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+                for i in 0..grid.width() {
+                    if grid.get(i, row_description.0) == CellState::Crossed {
+                        begin_offset += 1;
+                    }
+                    else {
+                        break;
+                    }
+                }
+
+                for i in (0..grid.width()).rev() {
```
Comment (alexesmet) [inline src/simple_engine.rs]: Can be rewritten using take while, rewrite or leave a TODO comment
```suggestion
                for i in (0..grid.width()).rev() { // TODO: rewrite using take-while
```

File: src/simple_engine.rs [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+                    if grid.get(i, row_description.0) == CellState::Crossed {
+                        end_offset += 1;
+                    }
+                    else {
+                        break;
+                    }
+                }
+
+                let total_filled_count = row_description.1.iter().map(|x| x.0).sum::<usize>();
+                let total_crossed_count = row_description.1.len() - 1;
```
Comment (alexesmet) [inline src/simple_engine.rs]:
```suggestion
                let minimum_crossed_count = row_description.1.len() - 1;
```

**Review by alexesmet — CHANGES_REQUESTED** [review]

**Pull request merged**